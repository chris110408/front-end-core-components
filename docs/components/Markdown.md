# Markdown

Markdown renderer, essentially a wrapper for [react-markdown](https://github.com/rexxars/react-markdown) providing a way of setting global default props via the [React Context API](https://reactjs.org/docs/context.html).

## Dependencies

* react `^16.3`
* prop-types `^15.6.2`
* react-markdown `^4.2.2`

## Reference

### Provider

| prop name | description | type | required? | default value |
| --------- | ----------- | ---- | --------- | ------------- |
| markdownProps | Props forwarded to [ReactMarkdown](https://github.com/rexxars/react-markdown) | `Array<Object>` | ✗ | - |
| children | Child nodes | `Node` | ✓ | - |

### Markdown

See [ReactMarkdown](https://github.com/rexxars/react-markdown) for available props. Will consume props from `Provider` where specified, any props defined directly on `Markdown` will overwrite props set via the `Provider`.

## Usage

### Provider (optional)

Put the provider high up in your application component tree, usually within `App.js` or the main entry point to your application/website.

```javascript
import React from 'react'
import { Provider } from '@input-output-hk/front-end-core-components/components/Markdown'

export default () => (
  <Provider markdownProps={{ ... }}>
    ...
    <p>The rest of your app</p>
    ...
  </Provider>
)

```

### Markdown

```javascript
import React from 'react'
import Markdown from '@input-output-hk/front-end-core-components/components/Markdown'

export default () => (
  <Markdown>
    {myMarkdown}
  </Markdown>
)

```
