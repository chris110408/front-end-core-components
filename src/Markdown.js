import React, { createContext } from 'react'
import PropTypes from 'prop-types'
import ReactMarkdown from 'react-markdown'

const MarkdownContext = createContext()

export const Provider = ({ markdownProps = {}, children }) => (
  <MarkdownContext.Provider value={{ markdownProps }}>
    {children}
  </MarkdownContext.Provider>
)

Provider.propTypes = {
  // Props used in ReactMarkdown component, can be overwritten at local component level
  // see https://github.com/rexxars/react-markdown
  markdownProps: PropTypes.object,
  children: PropTypes.node
}

const Markdown = (props) => (
  <MarkdownContext.Consumer>
    {({ markdownProps = {} } = {}) => (
      <ReactMarkdown
        {...markdownProps}
        {...props}
        renderers={{ ...markdownProps.renderers, ...props.renderers }}
        parserOptions={{ ...markdownProps.parserOptions, ...props.parserOptions }}
      />
    )}
  </MarkdownContext.Consumer>
)

Markdown.propTypes = {
  renderers: PropTypes.object,
  parserOptions: PropTypes.object
}

export default Markdown
