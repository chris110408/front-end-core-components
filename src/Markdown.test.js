import React from 'react'
import Markdown, { Provider } from './Markdown'
import { mount } from 'enzyme'

function getWrapper ({ props = {}, providerProps = {} } = {}) {
  return mount(
    <Provider {...providerProps}>
      <Markdown {...props} />
    </Provider>
  )
}

describe('<Markdown />', () => {
  test('it renders correctly with minimal props', () => {
    expect(getWrapper({ props: { source: '# A title' } })).toMatchSnapshot()
  })

  describe('setting props at a global level', () => {
    test('it renders using the global props', () => {
      expect(getWrapper({
        providerProps: { markdownProps: { source: '# A global title' } } })).toMatchSnapshot()
    })

    describe('overwriting the global props', () => {
      test('it renders with the local props overwriting the global props', () => {
        expect(getWrapper({
          providerProps: { markdownProps: { source: '# A global title' } },
          props: { source: '# A local title' }
        })).toMatchSnapshot()
      })
    })

    describe('setting renderers at a global level', () => {
      let GlobalTitle
      beforeEach(() => {
        GlobalTitle = (props) => (
          <h1 {...props} className='my-global-title' />
        )
      })

      test('it renders with the renderers at a global level', () => {
        expect(getWrapper({
          providerProps: { markdownProps: { renderers: { heading: GlobalTitle } } },
          props: { source: '# A custom title' }
        })).toMatchSnapshot()
      })

      describe('overwriting the global renderers', () => {
        let LocalTitle
        beforeEach(() => {
          LocalTitle = (props) => (
            <h1 {...props} className='my-local-title' />
          )
        })

        test('it renders with the local renderers overwriting the global renderers', () => {
          expect(getWrapper({
            providerProps: { markdownProps: { renderers: { heading: GlobalTitle } } },
            props: { source: '# Another custom title', renderers: { heading: LocalTitle } }
          })).toMatchSnapshot()
        })
      })
    })

    describe('setting parserOptions at a global level', () => {
      let globalParserOptions
      beforeEach(() => {
        globalParserOptions = { pedantic: true }
      })

      test('it renders with the renderers at a global level', () => {
        expect(getWrapper({
          providerProps: { markdownProps: { parserOptions: globalParserOptions } },
          props: { source: 'Some markdown' }
        })).toMatchSnapshot()
      })

      describe('overwriting the global renderers', () => {
        test('it renders with the local renderers overwriting the global renderers', () => {
          expect(getWrapper({
            providerProps: { markdownProps: { parserOptions: globalParserOptions } },
            props: { source: 'Some more markdown', parserOptions: { pedantic: false } }
          })).toMatchSnapshot()
        })
      })
    })
  })

  describe('when no Provider is present', () => {
    test('it still renders the markdown correctly', () => {
      expect(mount(<Markdown source='no global state' />)).toMatchSnapshot()
    })
  })
})
