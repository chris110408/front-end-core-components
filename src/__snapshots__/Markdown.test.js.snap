// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Markdown /> it renders correctly with minimal props 1`] = `
<Provider>
  <Markdown
    source="# A title"
  >
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="# A title"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <Heading
          key="heading-1-1-0"
          level={1}
        >
          <h1>
            <TextRenderer
              key="text-1-3-0"
              nodeKey="text-1-3-0"
              value="A title"
            >
              A title
            </TextRenderer>
          </h1>
        </Heading>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> setting props at a global level it renders using the global props 1`] = `
<Provider
  markdownProps={
    Object {
      "source": "# A global title",
    }
  }
>
  <Markdown>
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="# A global title"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <Heading
          key="heading-1-1-0"
          level={1}
        >
          <h1>
            <TextRenderer
              key="text-1-3-0"
              nodeKey="text-1-3-0"
              value="A global title"
            >
              A global title
            </TextRenderer>
          </h1>
        </Heading>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> setting props at a global level overwriting the global props it renders with the local props overwriting the global props 1`] = `
<Provider
  markdownProps={
    Object {
      "source": "# A global title",
    }
  }
>
  <Markdown
    source="# A local title"
  >
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="# A local title"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <Heading
          key="heading-1-1-0"
          level={1}
        >
          <h1>
            <TextRenderer
              key="text-1-3-0"
              nodeKey="text-1-3-0"
              value="A local title"
            >
              A local title
            </TextRenderer>
          </h1>
        </Heading>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> setting props at a global level setting parserOptions at a global level it renders with the renderers at a global level 1`] = `
<Provider
  markdownProps={
    Object {
      "parserOptions": Object {
        "pedantic": true,
      },
    }
  }
>
  <Markdown
    source="Some markdown"
  >
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={
        Object {
          "pedantic": true,
        }
      }
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="Some markdown"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <p
          key="paragraph-1-1-0"
        >
          <TextRenderer
            key="text-1-1-0"
            nodeKey="text-1-1-0"
            value="Some markdown"
          >
            Some markdown
          </TextRenderer>
        </p>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> setting props at a global level setting parserOptions at a global level overwriting the global renderers it renders with the local renderers overwriting the global renderers 1`] = `
<Provider
  markdownProps={
    Object {
      "parserOptions": Object {
        "pedantic": true,
      },
    }
  }
>
  <Markdown
    parserOptions={
      Object {
        "pedantic": false,
      }
    }
    source="Some more markdown"
  >
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={
        Object {
          "pedantic": false,
        }
      }
      plugins={Array []}
      rawSourcePos={false}
      renderers={Object {}}
      skipHtml={false}
      source="Some more markdown"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <p
          key="paragraph-1-1-0"
        >
          <TextRenderer
            key="text-1-1-0"
            nodeKey="text-1-1-0"
            value="Some more markdown"
          >
            Some more markdown
          </TextRenderer>
        </p>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> setting props at a global level setting renderers at a global level it renders with the renderers at a global level 1`] = `
<Provider
  markdownProps={
    Object {
      "renderers": Object {
        "heading": [Function],
      },
    }
  }
>
  <Markdown
    source="# A custom title"
  >
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={
        Object {
          "heading": [Function],
        }
      }
      skipHtml={false}
      source="# A custom title"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <GlobalTitle
          key="heading-1-1-0"
          level={1}
        >
          <h1
            className="my-global-title"
            level={1}
          >
            <TextRenderer
              key="text-1-3-0"
              nodeKey="text-1-3-0"
              value="A custom title"
            >
              A custom title
            </TextRenderer>
          </h1>
        </GlobalTitle>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> setting props at a global level setting renderers at a global level overwriting the global renderers it renders with the local renderers overwriting the global renderers 1`] = `
<Provider
  markdownProps={
    Object {
      "renderers": Object {
        "heading": [Function],
      },
    }
  }
>
  <Markdown
    renderers={
      Object {
        "heading": [Function],
      }
    }
    source="# Another custom title"
  >
    <ReactMarkdown
      astPlugins={Array []}
      escapeHtml={true}
      parserOptions={Object {}}
      plugins={Array []}
      rawSourcePos={false}
      renderers={
        Object {
          "heading": [Function],
        }
      }
      skipHtml={false}
      source="# Another custom title"
      sourcePos={false}
      transformLinkUri={[Function]}
    >
      <Root
        key="root-1-1-0"
      >
        <LocalTitle
          key="heading-1-1-0"
          level={1}
        >
          <h1
            className="my-local-title"
            level={1}
          >
            <TextRenderer
              key="text-1-3-0"
              nodeKey="text-1-3-0"
              value="Another custom title"
            >
              Another custom title
            </TextRenderer>
          </h1>
        </LocalTitle>
      </Root>
    </ReactMarkdown>
  </Markdown>
</Provider>
`;

exports[`<Markdown /> when no Provider is present it still renders the markdown correctly 1`] = `
<Markdown
  source="no global state"
>
  <ReactMarkdown
    astPlugins={Array []}
    escapeHtml={true}
    parserOptions={Object {}}
    plugins={Array []}
    rawSourcePos={false}
    renderers={Object {}}
    skipHtml={false}
    source="no global state"
    sourcePos={false}
    transformLinkUri={[Function]}
  >
    <Root
      key="root-1-1-0"
    >
      <p
        key="paragraph-1-1-0"
      >
        <TextRenderer
          key="text-1-1-0"
          nodeKey="text-1-1-0"
          value="no global state"
        >
          no global state
        </TextRenderer>
      </p>
    </Root>
  </ReactMarkdown>
</Markdown>
`;
